### PR TITLE
Ecdc 2447 handle motor msgtxt

### PR DESCRIPTION
**Pull request by kmurica**
More robust handling of motor states.

### CODE REVIEW

**Comment by kmurica** [comment]
After a couple of meetings and testing sessions with Torsten, I think this PR is ready for review.
One comment from my side: should we add the connection check more low level?
Additional work needs to be done for new messages that won't have a severity level (like power off, immovable etc) but this will be handled in a separate ticket by me and Torsten.

File: nicos_ess/devices/epics/motor.py [hunk 1]
```diff
@@ -59,6 +59,12 @@ class EpicsMotor(CanDisable, CanReference, HasOffset, EpicsAnalogMoveableEss,
         'reseterrorpv': Param('Optional PV with error reset switch.',
                               type=pvname, mandatory=False, settable=False,
                               userparam=False),
+        'errorseveritypv': Param('Optional PV with error severity.',
```
Comment (mattclarke) [inline nicos_ess/devices/epics/motor.py:62]: 'errorseveritypv'
Let's put some underscores in these names ;)

File: nicos_ess/devices/epics/motor.py [hunk 2]
```diff
@@ -234,12 +251,42 @@ def _get_status_message(self):
...
+
+    def _check_ioc_controller_communication(self):
+        """
+        Check if there is communication between the IOC and controller.
 
-        return self._get_pv('errormsgpv', as_string=True)
+        :return: returns communication error message if there is no connection,
+        otherwise returns an empty string.
+        """
+        _COMM_STAT = "COMM"
```
Comment (mattclarke) [inline nicos_ess/devices/epics/motor.py:277]: Why are these indicated as constants (uppercase) and private (leading _)?
They are local variables, right?
They probably should be extracted as constants though

File: nicos_ess/devices/epics/pva/monitor_motor.py [hunk 3]
```diff
@@ -252,12 +273,46 @@ def _get_status_message(self):
...
         """
         if not self.errormsgpv:
             return ''
+        error_msg = self._get_pv('errormsgpv', as_string=True)
+        if not self.errorseveritypv or not self.errorstatuspv:
+            return error_msg
+        error_severity = self._get_pv('errorseveritypv', as_string=True)
+        if error_severity:
+            error_status = self._get_pv('errorstatuspv', as_string=True)
+            return f"MSG: \"{error_msg}\", STATUS: {error_status}, " \
```
Comment (mattclarke) [inline nicos_ess/devices/epics/pva/monitor_motor.py:287]: If you use `'` for the outer quotes then you don't need to escape the inner ones, e.g. `'this is "quoted"'`

File: nicos_ess/devices/epics/pva/monitor_motor.py [hunk 4]
```diff
@@ -252,12 +273,46 @@ def _get_status_message(self):
...
+        :return: returns communication error message if there is no connection,
+        otherwise returns an empty string.
+        """
+        _COMM_STAT = "COMM"
+        _INVALID_SEVR = "INVALID"
+        if not self.errorseveritypv or not self.errorstatuspv:
+            return ""
+        error_severity = self._get_pv('errorseveritypv', as_string=True)
+        if error_severity == _INVALID_SEVR:
+            error_status = self._get_pv('errorstatuspv', as_string=True)
```
Comment (mattclarke) [inline nicos_ess/devices/epics/pva/monitor_motor.py:305]: There is a lot of code here that looks duplicated.

File: nicos_ess/devices/epics/motor.py [hunk 5]
```diff
@@ -234,12 +251,42 @@ def _get_status_message(self):
...
         """
         if not self.errormsgpv:
             return ''
+        error_msg = self._get_pv('errormsgpv', as_string=True)
+        if not self.errorseveritypv or not self.errorstatuspv:
+            return error_msg
+        error_severity = self._get_pv('errorseveritypv', as_string=True)
+        if error_severity:
+            error_status = self._get_pv('errorstatuspv', as_string=True)
+            return f"MSG: \"{error_msg}\", STATUS: {error_status}, " \
```
Comment (mattclarke) [inline nicos_ess/devices/epics/motor.py:265]: There is a lot of information in this message, a lot of it doesn't seem relevant to the user.
Just the error message should be sufficient, the rest can go in the log.

File: nicos_ess/devices/epics/motor.py (hunk 2 above)
Comment (kmurica) [reply nicos_ess/devices/epics/motor.py]: Yup, I left it a bit ambivalent because I was not sure if I should move everything elsewhere completely.
But if it is fine to have it in the device class, I will keep it there. Moved the variables to class attributes.

File: nicos_ess/devices/epics/motor.py (hunk 5 above)
Comment (kmurica) [reply nicos_ess/devices/epics/motor.py]: Yes, you are right, probably the severity and status is only interesting to us and the EPICS experts. Moved most of it to log. Sending only error msg text to user.

File: nicos_ess/ymir/setups/alignment_stack.py [hunk 6]
```diff
@@ -10,6 +10,8 @@
         errormsgpv='SES-SCAN:MC-MCU-001:m1-MsgTxt',
         errorbitpv='SES-SCAN:MC-MCU-001:m1-Err',
         reseterrorpv='SES-SCAN:MC-MCU-001:m1-ErrRst',
+        error_severity_pv='SES-SCAN:MC-MCU-001:m1.SEVR',
+        error_status_pv='SES-SCAN:MC-MCU-001:m1.STAT',
```
Comment (mattclarke) [inline nicos_ess/ymir/setups/alignment_stack.py:14]: These two aren't needed as SEVR and STAT are fields not PVs, so they are always present by default.
It is just we are not currently using them, but that can be fixed by adding them to `_record_fields` or the `def _get_record_fields(self):` depending on which version of the motor.

Let me know if you need a refresher on fields vs PVs or anything else in this comment that is unclear :D

File: nicos_ess/devices/epics/pva/monitor_motor.py [hunk 7]
```diff
@@ -252,12 +284,64 @@ def _get_status_message(self):
...
+                self._get_pv('error_status_pv', as_string=True)
+        return epics_msg_pvs
+
+    def _get_epics_err_info(self, error):
+        """
+        Sends EPICS error information to the NICOS logger.
+
+        :return: returns an error message string.
+        """
+        session.log.error([f'EPICS ERROR MESSAGE: '
```
Comment (mattclarke) [inline nicos_ess/devices/epics/pva/monitor_motor.py:338]:
```suggestion
        self.log.error([f'EPICS ERROR MESSAGE: '
```
Then you don't need to import session

File: nicos_ess/devices/epics/pva/monitor_motor.py [hunk 8]
```diff
@@ -252,12 +284,64 @@ def _get_status_message(self):
...
+        """
+        epics_msg_pvs = self._parse_epics_message_pvs()
+        if epics_msg_pvs[self.SEVR] == self.INVALID_SEVR and \
+                epics_msg_pvs[self.STAT] == self.COMM_STAT:
+            return self._get_epics_err_info(epics_msg_pvs)
+        else:
             return ''
 
-        return self._get_pv('errormsgpv', as_string=True)
+    def _parse_epics_message_pvs(self):
```
Comment (mattclarke) [inline nicos_ess/devices/epics/pva/monitor_motor.py:310]: It is not "parsing" it is just reading them

File: nicos_ess/devices/epics/pva/monitor_motor.py [hunk 9]
```diff
@@ -252,12 +284,64 @@ def _get_status_message(self):
...
+                         self.STAT: "",
+                         self.MSG_TXT: ""}
+        if self.errormsgpv:
+            epics_msg_pvs[self.MSG_TXT] = \
+                self._get_pv('errormsgpv', as_string=True)
+        if self.error_severity_pv:
+            epics_msg_pvs[self.SEVR] = \
+                self._get_pv('error_severity_pv', as_string=True)
+        if self.error_status_pv:
+            epics_msg_pvs[self.STAT] = \
```
Comment (mattclarke) [inline nicos_ess/devices/epics/pva/monitor_motor.py:328]: These two ifs can be removed when SEVR and STAT are fields

File: nicos_ess/devices/epics/pva/monitor_motor.py [hunk 10]
```diff
@@ -105,6 +113,16 @@ class EpicsMotor(CanDisable, CanReference, HasOffset, EpicsMoveable, Motor):
         'writepv': 'target',
     }
 
+    # Severity and message status when connection between IOC and controller is
+    # lost.
+    COMM_STAT = "COMM"
+    INVALID_SEVR = "INVALID"
```
Comment (mattclarke) [inline nicos_ess/devices/epics/pva/monitor_motor.py:119]: These two values are only used once, so they might as well be inlined

File: nicos_ess/ymir/setups/alignment_stack.py (hunk 6 above)
Comment (kmurica) [reply nicos_ess/ymir/setups/alignment_stack.py:14]: Thanks for the clarification here. Completely forgot that about fields! :)

File: nicos_ess/devices/epics/pva/monitor_motor.py [hunk 11]
```diff
@@ -252,12 +284,64 @@ def _get_status_message(self):
...
+        if self.errormsgpv:
+            epics_msg_pvs[self.MSG_TXT] = \
+                self._get_pv('errormsgpv', as_string=True)
+        if self.error_severity_pv:
+            epics_msg_pvs[self.SEVR] = \
+                self._get_pv('error_severity_pv', as_string=True)
+        if self.error_status_pv:
+            epics_msg_pvs[self.STAT] = \
+                self._get_pv('error_status_pv', as_string=True)
+        return epics_msg_pvs
```
Comment (mattclarke) [inline nicos_ess/devices/epics/pva/monitor_motor.py:330]: It's probably not clear at the moment, but once the other bits are tidied up this can just return a tuple (probably)

Comment (mattclarke) [reply nicos_ess/devices/epics/pva/monitor_motor.py:330]: In which case, the dictionary keys can be deleted

File: nicos_ess/devices/epics/pva/monitor_motor.py [hunk 12]
```diff
@@ -252,12 +284,64 @@ def _get_status_message(self):
...
+
+    def _get_epics_err_info(self, error):
+        """
+        Sends EPICS error information to the NICOS logger.
+
+        :return: returns an error message string.
+        """
+        session.log.error([f'EPICS ERROR MESSAGE: '
+                           f'"{error[self.MSG_TXT]}"',
+                           f'EPICS ERROR MESSAGE STATUS:'
```
Comment (mattclarke) [inline nicos_ess/devices/epics/pva/monitor_motor.py:340]: No need for shouting ;)
Perhaps can be something like "Motor error: {reason} ({severity}, {status})"
Check with Torsten but I think the severity can be none, minor, major, invalid. Perhaps for minor we can use warning rather than error?

**Comment by mattclarke** [comment]
This is looking promise (despite all the comments).
I am a bit worried that YMIR is showing this:
`ERROR [2021-04-23 11:42:16] ['EPICS ERROR MESSAGE: "Moving", blah blah`
I don't think moving should be considered an error ;)

File: nicos_ess/devices/epics/pva/monitor_motor.py (hunk 11 above)
Comment (kmurica) [reply nicos_ess/devices/epics/pva/monitor_motor.py:311]: Would it be a problem to keep the dictionary for now? As this is still work in progress together with Torsten and I am not sure how additional messages will be handled that now lack a proper severity level and status (still to be defined by Torsten if there will be a separate PV or severity). Once this is completed with the other tickets, I can refactor it further and get rid of the dictionary if needed.

**Comment by kmurica** [comment]
> This is looking promise (despite all the comments).
> I am a bit worried that YMIR is showing this:
> `ERROR [2021-04-23 11:42:16] ['EPICS ERROR MESSAGE: "Moving", blah blah`
> I don't think moving should be considered an error ;)

@mattclarke Tried to replicate this but it looks normal (yellow when moving and no other errors are present). Is this still an issue that you can see?

File: nicos_ess/devices/epics/pva/monitor_motor.py (hunk 11 above)
Comment (kmurica) [reply nicos_ess/devices/epics/pva/monitor_motor.py:301]: Changed from dictionary to tuple.

File: nicos_ess/devices/epics/pva/monitor_motor.py [hunk 13]
```diff
@@ -219,9 +221,16 @@ def doReadTarget(self):
...
 
     def doStatus(self, maxage=0):
-        stat, message = EpicsMoveable.doStatus(self)
+        # Before anything, check IOC and controller communication status.
+        comm_fail_stat = self._check_ioc_controller_communication()
+        if comm_fail_stat:
+            return status.ERROR, comm_fail_stat
 
-        if stat in [status.WARN, status.ERROR]:
+        stat, message = EpicsMoveable.doStatus(self)
```
Comment (mattclarke) [inline nicos_ess/devices/epics/pva/monitor_motor.py:229]: This only applies to monitor motor, but with some refactoring we can eliminate this call to the base class which essentially does the same thing as `read_epics_message_pvs` does.

File: nicos_ess/devices/epics/pva/monitor_motor.py [hunk 14]
```diff
@@ -252,12 +255,61 @@ def _get_status_message(self):
...
+                     f'{error_status})'
+        if stat == status.WARN:
+            self.log.warning(msg_to_log)
+            return f'Motor warning: "{msg_txt}"', status.WARN
+        elif stat == status.ERROR:
+            self.log.error(msg_to_log)
+            return f'Motor error: "{msg_txt}"', status.ERROR
+
+    def _get_message_status(self, error_severity, error_status):
+        if error_severity == 'INVALID' and error_status == 'COMM':
```
Comment (mattclarke) [inline nicos_ess/devices/epics/pva/monitor_motor.py:303]: There is already a conversion in NICOS:
```
SEVERITY_TO_STATUS = {
    0: status.OK,  # NO_ALARM
    1: status.WARN,  # MINOR
    2: status.ERROR,  # MAJOR
    3: status.UNKNOWN,  # INVALID
}```

Comment (mattclarke) [reply nicos_ess/devices/epics/pva/monitor_motor.py:303]: But what you have done here might be better.
I'll need to think on it for a bit

**Comment by kmurica** [comment]
@mattclarke merged. I will keep the branch alive as agreed.
